app/controllers/users_controller.rb

```diff
@@ -496,28 +498,61 @@ def is_local_username
     usernames -= groups
     usernames.each(&:downcase!)
 
-    cannot_see = []
+    users = User
+      .where(staged: false)
+      .where(username_lower: usernames)
+      .index_by(&:username_lower)
+
+    cannot_see = {}
     here_count = nil
 
     topic_id = params[:topic_id]
     if topic_id.present? && topic = Topic.find_by(id: topic_id)
+      topic_muted_by = TopicUser
+        .where(topic: topic)
+        .where(user_id: users.values.map(&:id))
+        .where(notification_level: TopicUser.notification_levels[:muted])
+        .pluck(:user_id)
+        .to_set
+
+      topic_allowed_user_ids = TopicAllowedUser
+        .where(topic: topic)
+        .where(user_id: users.values.map(&:id))
+        .pluck(:user_id)
+        .to_set
+
+      topic_allowed_group_ids = TopicAllowedGroup
+        .where(topic: topic)
+        .pluck(:group_id)
+        .to_set
+
       usernames.each do |username|
-        if !Guardian.new(User.find_by_username(username)).can_see?(topic)
-          cannot_see.push(username)
+        user = users[username]
+        next if user.blank?
+
+        cannot_see_reason = nil
+        if !user.guardian.can_see?(topic)
+          cannot_see_reason = topic.private_message? ? :private : :category
+        elsif topic_muted_by.include?(user.id)
+          cannot_see_reason = :muted_topic
+        elsif topic.private_message? && !topic_allowed_user_ids.include?(user.id) && !user.group_ids.any? { |group_id| topic_allowed_group_ids.include?(group_id) }
```

Comment:
Sorry can we write a test for this because I'm trying to work through the conditions here and I don't think this condition will ever be true since it would mean that the user cannot view the topic. If the user cannot view the topic, `!user.guardian.can_see?(topic)` will evaluate to true and skip this condition totally.

Reply:
Sure, I added a test.